From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nick Chan <towinchenmi@gmail.com>
Date: Sat, 7 Feb 2026 01:22:03 +0800
Subject: [PATCH 18/21] mfd: macsmc: Add T8012 and T8015 SMC support

The SMC on T8012 and T8015 SoCs works in the same way. However, there
are differences in reboot and RTC handling that cannot be probed at
runtime. Add support for these SMCs by declaring a different list of
subdevices for them.

Signed-off-by: Nick Chan <towinchenmi@gmail.com>
---
 drivers/mfd/macsmc.c | 46 +++++++++++++++++++++++++++++++++++++++-----
 1 file changed, 41 insertions(+), 5 deletions(-)

diff --git a/drivers/mfd/macsmc.c b/drivers/mfd/macsmc.c
index 5c71e7d36..08f7b6c8e 100644
--- a/drivers/mfd/macsmc.c
+++ b/drivers/mfd/macsmc.c
@@ -44,7 +44,19 @@
 
 #define SMC_TIMEOUT_MS		500
 
-static const struct mfd_cell apple_smc_devs[] = {
+#define SMC_TYPE_T8012		0x8012
+#define SMC_TYPE_T8015		0x8015
+#define SMC_TYPE_T8103		0x8103
+
+static const struct mfd_cell apple_smc_t8015_devs[] = {
+	MFD_CELL_NAME("macsmc-input"),
+	MFD_CELL_NAME("macsmc-power"),
+	MFD_CELL_OF("macsmc-gpio", NULL, NULL, 0, 0, "apple,smc-gpio"),
+	MFD_CELL_OF("macsmc-hwmon", NULL, NULL, 0, 0, "apple,smc-hwmon"),
+	MFD_CELL_OF("macsmc-reboot", NULL, NULL, 0, 0, "apple,t8015-smc-reboot"),
+};
+
+static const struct mfd_cell apple_smc_t8103_devs[] = {
 	MFD_CELL_NAME("macsmc-input"),
 	MFD_CELL_NAME("macsmc-power"),
 	MFD_CELL_OF("macsmc-gpio", NULL, NULL, 0, 0, "apple,smc-gpio"),
@@ -411,7 +423,30 @@ static int apple_smc_probe(struct platform_device *pdev)
 	struct device *dev = &pdev->dev;
 	struct apple_smc *smc;
 	__be32 count;
+	u32 dev_cnt;
+	const struct mfd_cell *devs;
 	int ret;
+	uintptr_t type;
+
+	type = (uintptr_t)of_device_get_match_data(dev);
+	switch (type) {
+		/*
+		 * T8012 has differences in button and RTC handling
+		 * Currently it works with T8015.
+		 */
+		case SMC_TYPE_T8012:
+		case SMC_TYPE_T8015:
+			devs = apple_smc_t8015_devs;
+			dev_cnt = ARRAY_SIZE(apple_smc_t8015_devs);
+			break;
+		case SMC_TYPE_T8103:
+			devs = apple_smc_t8103_devs;
+			dev_cnt = ARRAY_SIZE(apple_smc_t8103_devs);
+			break;
+		default:
+			WARN_ON(1);
+			return -ENODEV;
+	}
 
 	smc = devm_kzalloc(dev, sizeof(*smc), GFP_KERNEL);
 	if (!smc)
@@ -478,8 +513,7 @@ static int apple_smc_probe(struct platform_device *pdev)
 	if (ret)
 		return ret;
 
-	ret = devm_mfd_add_devices(smc->dev, PLATFORM_DEVID_NONE,
-				   apple_smc_devs, ARRAY_SIZE(apple_smc_devs),
+	ret = devm_mfd_add_devices(smc->dev, PLATFORM_DEVID_NONE, devs, dev_cnt,
 				   NULL, 0, NULL);
 	if (ret)
 		return dev_err_probe(smc->dev, ret, "Failed to register sub-devices");
@@ -489,8 +523,10 @@ static int apple_smc_probe(struct platform_device *pdev)
 }
 
 static const struct of_device_id apple_smc_of_match[] = {
-	{ .compatible = "apple,t8103-smc" },
-	{ .compatible = "apple,smc" },
+	{ .compatible = "apple,t8012-smc", .data = (void*)SMC_TYPE_T8012 },
+	{ .compatible = "apple,t8015-smc", .data = (void*)SMC_TYPE_T8015 },
+	{ .compatible = "apple,t8103-smc", .data = (void*)SMC_TYPE_T8103 },
+	{ .compatible = "apple,smc", .data = (void*)SMC_TYPE_T8103 },
 	{},
 };
 MODULE_DEVICE_TABLE(of, apple_smc_of_match);
